vm_operator/src/charm.py

```diff
@@ -0,0 +1,263 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical
+# See LICENSE file for licensing details.
+
+"""Ubuntu Software Centre ratings service.
+
+A backend service to support application ratings in the new Ubuntu Software Centre.
+"""
+
+import logging
+import secrets
+
+import ops
+
+from pathlib import Path
+from git import Repo
+import shutil
+from charms.operator_libs_linux.v0 import apt, systemd
+from charms.data_platform_libs.v0.data_interfaces import DatabaseCreatedEvent, DatabaseRequires
+from ops.model import MaintenanceStatus, BlockedStatus, ActiveStatus
+from ops.framework import StoredState
+from subprocess import check_output, CalledProcessError
+import os
+from os import environ
+from jinja2 import Template
+
+logger = logging.getLogger(__name__)
+
+
+APP_PATH = Path("/srv/app")
+UNIT_PATH = Path("/etc/systemd/system/ratings.service")
+CARGO_PATH = Path(environ.get("HOME", "/root")) / ".cargo/bin/cargo"
+
+
+class RatingsCharm(ops.CharmBase):
+    """Main operator class for ratings service."""
+
+    _stored = StoredState()
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self._ratings_svc = None
+
+        # Initialise the integration with PostgreSQL
+        self._database = DatabaseRequires(self, relation_name="database", database_name="ratings")
+        self.framework.observe(self._database.on.database_created, self._on_database_created)
+        self.framework.observe(self.on.install, self._on_install)
+        self._stored.set_default(repo="", port="", conn_str="", install_completed=False)
+        self.framework.observe(self.on.start, self._on_start)
+        self.framework.observe(self.on.pull_and_rebuild_action, self._on_pull_and_rebuild)
+
+    def _on_start(self, _):
+        """Start the workload."""
+        # Enable and start the "ratings" systemd unit
+        systemd.service_resume("ratings")
+        self.unit.status = ActiveStatus()
+
+    def _on_install(self, _):
+        """Install prerequisites for the application."""
+        self.unit.status = MaintenanceStatus("Installing rustc, cargo and other dependencies")
+
+        # Install via apt
+        self._install_apt_packages(["curl", "git", "gcc", "libssl-dev", "pkg-config","protobuf-compiler"])
+
+        # Ensure squid proxy, done after apt to not interfere
+        self._set_squid_proxy()
+
+        # Curl minial rust toolchain
+        try:
+            check_output("curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -y --profile minimal", shell=True)
+            self._stored.install_completed = True
+            self.unit.status = MaintenanceStatus("Installation complete, waiting for database.")
+
+        except CalledProcessError as e:
+            logger.error(f"Curl command failed with error code {e.returncode}")
+            self.unit.status = BlockedStatus("Curl command failed")
+            return
```

Comment:
This is pretty heavy, and you'll end up with the whole toolchain installed on the machines running in production.

If you're planning on running on VMs in the long term, I'd recommend building this as a snap, then installing the snap package (pinned to a specific revision) in your charm. FWIW you can use the same `parts` definition from the removed `rockcraft.yaml` to build the snap :)